src/System/Random/Internal.hs

```diff
@@ -0,0 +1,1040 @@
+-- multiplication method.
+--
+-- Daniel Lemire. 2019. Fast Random Integer Generation in an Interval. In ACM
+-- Transactions on Modeling and Computer Simulation
+-- https://doi.org/10.1145/3230636
+--
+-- PRECONDITION (unchecked): s > 0
+boundedExclusiveIntegralM :: forall a g m . (Bits a, Integral a, StatefulGen g m) => a -> g -> m a
+boundedExclusiveIntegralM (s :: a) gen = go
+  where
+    n = integralWordSize s
+    -- We renamed 'L' from the paper to 'k' here because 'L' is not a valid
+    -- variable name in Haskell and 'l' is already used in the algorithm.
+    k = wordSizeInBits * n
+    twoToK = (1::a) `shiftL` k
+    modTwoToKMask = twoToK - 1
+
+    t = (twoToK - s) `mod` s
+    go :: (Bits a, Integral a, StatefulGen g m) => m a
+    go = do
+      x <- uniformIntegralWords n gen
+      let m = x * s
+      -- m .&. modTwoToKMask == m `mod` twoToK
+      let l = m .&. modTwoToKMask
+      if l < t
+        then go
+        -- m `shiftR` k == m `quot` twoToK
+        else return $ m `shiftR` k
+{-# INLINE boundedExclusiveIntegralM #-}
+
+-- | @integralWordSize i@ returns that least @w@ such that
+-- @i <= WORD_SIZE_IN_BITS^w@.
+integralWordSize :: (Bits a, Num a) => a -> Int
+integralWordSize = go 0
+  where
+    go !acc i
+      | i == 0 = acc
+      | otherwise = go (acc + 1) (i `shiftR` wordSizeInBits)
+{-# INLINE integralWordSize #-}
+
+-- | @uniformIntegralWords n@ is a uniformly pseudo-random integral in the range
+-- @[0, WORD_SIZE_IN_BITS^n)@.
+uniformIntegralWords :: (Bits a, Integral a, StatefulGen g m) => Int -> g -> m a
+uniformIntegralWords n gen = go 0 n
+  where
+    go !acc i
+      | i == 0 = return acc
+      | otherwise = do
+        (w :: Word) <- uniformM gen
+        go ((acc `shiftL` wordSizeInBits) .|. fromIntegral w) (i - 1)
+{-# INLINE uniformIntegralWords #-}
+
+-- | Uniformly generate an 'Integral' in an inclusive-inclusive range.
+--
+-- Only use for integrals size less than or equal to that of 'Word32'.
+unbiasedWordMult32RM :: (StatefulGen g m, Integral a) => (a, a) -> g -> m a
+unbiasedWordMult32RM (b, t) g
+  | b <= t    = (+b) . fromIntegral <$> unbiasedWordMult32 (fromIntegral (t - b)) g
+  | otherwise = (+t) . fromIntegral <$> unbiasedWordMult32 (fromIntegral (b - t)) g
+{-# SPECIALIZE unbiasedWordMult32RM :: StatefulGen g m => (Word8, Word8) -> g -> m Word8 #-}
+
+-- | Uniformly generate Word32 in @[0, s]@.
+unbiasedWordMult32 :: StatefulGen g m => Word32 -> g -> m Word32
+unbiasedWordMult32 s g
+  | s == maxBound = uniformWord32 g
+  | otherwise = unbiasedWordMult32Exclusive (s+1) g
+{-# INLINE unbiasedWordMult32 #-}
+
+-- | See [Lemire's paper](https://arxiv.org/pdf/1805.10941.pdf),
+-- [O\'Neill's
+-- blogpost](https://www.pcg-random.org/posts/bounded-rands.html) and
+-- more directly [O\'Neill's github
+-- repo](https://github.com/imneme/bounded-rands/blob/3d71f53c975b1e5b29f2f3b05a74e26dab9c3d84/bounded32.cpp#L234).
+-- N.B. The range is [0,t) **not** [0,t].
```

Comment:
Should probably be `N.B. The range is [0,r) **not** [0,r].`

Reply:
Make sense. Changed.